src/sptps_test.c

```diff
@@ -125,6 +134,149 @@ static void usage() {
 	fprintf(stderr, "Report bugs to tinc@tinc-vpn.org.\n");
 }
 
+#ifdef HAVE_MINGW
+
+// Windows does not allow calling select() on anything but sockets. Therefore,
+// to keep the same code as on other operating systems, we have to put a
+// separate thread between the stdin and the sptps loop way below. This thread
+// reads stdin and sends its content to the main thread through a TCP socket,
+// which can be properly select()'ed.
+void *stdin_reader_thread(void *arg) {
+	int *sock_fd = arg;
+	struct sockaddr_in sa;
+	socklen_t sa_size = sizeof(sa);
+
+	while(true) {
+		int peer_fd = accept(*sock_fd, (struct sockaddr *) &sa, &sa_size);
+
+		if(peer_fd < 0) {
+			fprintf(stderr, "accept() failed: %s\n", strerror(errno));
+			continue;
+		}
+
+		if(verbose) {
+			fprintf(stderr, "New connection received from :%d\n", ntohs(sa.sin_port));
+		}
+
+		char buf[1024];
+		ssize_t nread;
+
+		while((nread = read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
+			if(verbose) {
+				fprintf(stderr, "Read %lld bytes from input\n", nread);
+			}
+
+			ssize_t nsend = send(peer_fd, buf, nread, 0);
+
+			if(nsend <= 0) {
```

Comment:
`send()` might send less bytes than you request. You have to handle this by adding a loop around it that send the remaining bytes if necessary. Also, if it returns -1, then this might still not be a fatal error. Use:

```
if (nsend == -1) {
    if (sockwouldblock(sockerrno)) {
        /* retry */
    } else {
        /* it really is an error, abort */
    }
}
```